adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1718,38 +1799,73 @@ private boolean stopCalled() {
         }
 
         /**
-         * Closes a connection to a client.
+         * Actively closes the connection to the device.
+         *
+         * @param reason The reason for closing the connection.
+         * @param sendDisconnectedEvent {@code true} if events shall be sent concerning the disconnect.
+         * @throws NullPointerException if reason is {@code null}.
+         */
+        protected final void close(final String reason, final boolean sendDisconnectedEvent) {
+            Objects.requireNonNull(reason);
+            endpoint.closeHandler(v -> {});
+            onCloseInternal("close device connection", reason, sendDisconnectedEvent);
+        }
+
+        /**
+         * To be called by the MqttEndpoint closeHandler.
          */
         protected final void onClose() {
-            final String operationName = stopCalled() ? "CLOSE on server shutdown" : "CLOSE";
-            final Span span = newSpan(operationName);
+            if (stopCalled()) {
+                onCloseInternal("close device connection (server shutdown)", "server shutdown", true);
+            } else {
+                onCloseInternal("CLOSE", null, true);
+            }
+        }
+
+        private void onCloseInternal(final String spanOperationName, final String reason, final boolean sendDisconnectedEvent) {
```

Comment:
It seems to me that the invoking method should be in charge of the Span's lifecycle. So FMPOV we should rather accept a Span instead of an operation name and return a Future so that the invoking method can then finish the span.

Reply:
I've adapted the methods accordingly.